content/blog/2025-03-06-path-profiling.md

```diff
@@ -0,0 +1,98 @@
+
+Of course there was not really space to do a deeper analysis on edge vs. path profiling in this paper. Which is likely why that analysis was performed separately in the Ball paper from the background section. Although, disappointingly, even that paper doesn't include any experiments on actually optimizing programs with path and edge profiles.
+
+### Role in History
+
+This paper was the first to offer a performant algorithm for path profiling, so almost all of the other work on path profiling uses the paper's algorithm. The question here then just rounds to asking "what is the role of path profiling in history?"
+
+The authors mention that path profiling could be potentially used for performance tuning, profile-directed compilation, and test coverage. For test coverage, it doesn't really see any use in practice for the reasons mentioned in [the discussion](https://github.com/sampsyo/cs6120/discussions/487) (too much performance cost for not enough benefit over edge profiling). For performance tuning and profile-directed compilation, it seems like path profiling sees some use in non-real time applications, but I'm not sure how often they're used in practice. Interestingly, profile-guided optimization tools like [these](https://research.facebook.com/file/900986544473313/Improved-Basic-Block-Reordering.pdf) [two](https://arxiv.org/abs/1810.00905) block layout optimizers that might be able to benefit from path profiling don't use it. Optimistically, this suggests that path profiling offers more utility than block or edge profiling, but that it can be difficult to make use of that utility. Pessimistically, this suggests that path profiling doesn't offer much more than block or edge profiling. The Codestitcher authors do mention that the information path profiling offers is "excessive" for code layout optimization, which I guess counts towards both viewpoints depending on how you want to think about it. For similar reasons as in profile-guided optimization, path profiling is not used in practice for real-time use cases like just-in-time compilation. Most of them also rely on simpler techniques like counting function calls or branch paths in order to determine when to patch in code. However, some other reasons why path profiling might be difficult to use here are the 30% performance drop and the difficulty of effective interprocedural optimization.
```

Comment:
I recommend toning down your claims here along the lines of "path profiling is not used in practice for $PURPOSE". It would be more reasonable to say, "we looked around at a bunch of papers, and we couldn't find anyone saying they used path profiling." It would be good to humbly acknowledge that your searches are not somehow an exhaustive survey.